test/unit/SmartContracts/state.unit.test.mjs

```diff
@@ -0,0 +1,772 @@
+    expect((await state.getProposer(addr1.address)).thisAddress).to.equal(
+      ethers.constants.AddressZero,
+    );
+
+    await state.setProposer(addr1.address, [
+      addr1.address,
+      addr1.address,
+      addr2.address,
+      newUrl,
+      newFee,
+      false,
+      0,
+    ]);
+
+    expect((await state.getProposer(addr1.address)).thisAddress).to.equal(addr1.address);
+    expect((await state.proposers(addr1.address)).thisAddress).to.equal(addr1.address);
+    expect((await state.proposers(addr1.address)).nextAddress).to.equal(addr2.address);
+    expect((await state.proposers(addr1.address)).url).to.equal(newUrl);
+    expect((await state.proposers(addr1.address)).fee).to.equal(newFee);
+
+    await state.setProposer(addr2.address, [
+      addr2.address,
+      addr2.address,
+      addr2.address,
+      newUrl,
+      newFee,
+      false,
+      0,
+    ]);
+    await state.setNumProposers(2);
+    await state.setStakeAccount(addr1.address, amount, challengeLocked);
+    await state.setStakeAccount(addr2.address, amount, challengeLocked);
+
+    await state.setCurrentProposer(addr1.address);
+
+    expect((await state.getCurrentProposer()).thisAddress).to.equal(addr1.address);
+
+    await state.changeCurrentProposer();
+
+    expect((await state.getCurrentProposer()).thisAddress).to.equal(addr2.address);
+  });
+
+  it('should not change current proposer with numProposers > 1 and 49 wei in stake', async function () {
+    const newUrl = 'url';
+    const newFee = 100;
+    const amount = 49;
+    const challengeLocked = 5;
+
+    expect((await state.getCurrentProposer()).thisAddress).to.equal(ethers.constants.AddressZero);
+    expect((await state.getProposer(addr1.address)).thisAddress).to.equal(
+      ethers.constants.AddressZero,
+    );
+
+    await state.setProposer(addr1.address, [
+      addr1.address,
+      addr1.address,
+      addr2.address,
+      newUrl,
+      newFee,
+      false,
+      0,
+    ]);
+
+    expect((await state.getProposer(addr1.address)).thisAddress).to.equal(addr1.address);
+    expect((await state.proposers(addr1.address)).thisAddress).to.equal(addr1.address);
+    expect((await state.proposers(addr1.address)).nextAddress).to.equal(addr2.address);
+    expect((await state.proposers(addr1.address)).url).to.equal(newUrl);
+    expect((await state.proposers(addr1.address)).fee).to.equal(newFee);
+
+    await state.setProposer(addr2.address, [
+      addr2.address,
+      addr2.address,
+      addr2.address,
+      newUrl,
+      newFee,
+      false,
+      0,
+    ]);
+    await state.setNumProposers(2);
+    await state.setStakeAccount(addr1.address, amount, challengeLocked);
+    await state.setStakeAccount(addr2.address, amount, challengeLocked);
+    await state.setCurrentProposer(addr1.address);
+
+    expect((await state.getCurrentProposer()).thisAddress).to.equal(addr1.address);
+
+    await expect(state.changeCurrentProposer()).to.be.reverted;
```

Comment:
I am not sure if this test makes sense.
I included it because maybe it might make sense to include a `require`.
the revert occurs in the `spanProposerSet` function at line `664` when it tries to access the `slots` array.
it tries to access the array by iterating a number of times equal to `proposerSetCount` (which is set to 5).
               
while the number of slots is calculated like this: 
`weight = stake.amount / valuePerSlot;` (with `valuePerSlot` set to 10)

then we could enter a `require` where the stake account must be equal to or greater than 
`proposerSetCount* valuePerSlot`

otherwise there will be fewer slots generated than expected